### ISSUE TITLE
Appimage Commandline Notification of .config and .home interferes with output

### ISSUE THREAD

**Issue by MerryMurray** [AppImage/AppImageKit#923]
HI
Newbie at creating appimages. Created a bash appimage to be run in the terminal.
Using .config and .home folders.
When run get notification in terminal the .config and .home are being used.
Notification is sent to stdout not stderr interfering with appimage command output.
Notification is displayed on every run.
Could it be changed to use stderr and just once?
User having to do something like 'dmesg | myappimage 2>/dev/null' is awkward?
Asked at IRC to file a ticket but not sure how?

Thanks

Murray

**Comment by TheAssassin** [comment]
Printing on stderr is not a problem at all, will fix that in a couple of minutes.

"Printing once", well, how would we remember we printed the message(s) already? There's no good way I could come up with. We'd have to e.g., create some hidden file in the two directories, but that's pretty ugly IMO. Using access times is even worse (they're sometimes even disabled on filesystems) Maybe @probonopd or @azubieta have a better idea?

**Comment by MerryMurray** [comment]
Thanks for the reply.  Apologies for the late reply.Glad the change to stederr is doable.RE Just Once messages
  agreed keeping hidden files is ugly and fragile. Looking for a solution -  maybe modifying an AppImage file attribute but couldn't find one.  how about not displaying the message at all.  Use something like  --image-status to display them.  Most appimages are gui so few will  see the message anyway. It is sad for the user to have to do:    dmesg | myappimage >2>dev/null   everywhere.
Impressed with appimages . Used so very littlte space. Have two bashappimages 5.8 and 6.0 mb including a text editor.  Well done. Thanks
Murray
  
===================================================================================

    On Saturday, February 9, 2019, 10:46:06 p.m. PST, TheAssassin <notifications@github.com> wrote:  
 
 
Printing on stderr is not a problem at all, will fix that in a couple of minutes.

"Printing once", well, how would we remember we printed the message(s) already? There's no good way I could come up with. We'd have to e.g., create some hidden file in the two directories, but that's pretty ugly IMO. Using access times is even worse (they're sometimes even disabled on filesystems) Maybe @probonopd or @azubieta have a better idea?

**Comment by probonopd** [comment]
> Using .config and .home folders.

That does not make sense imho, only __one__ of the two should be used.

**Comment by probonopd** [comment]
> Maybe @probonopd or @azubieta have a better idea?

It should be printed every time in my opinion, so that users know exactly what is going on. If people think it should be printed to STRERR, I am indifferent to that (was assuming that STDERR was for _errors_ but appartenly it is for messages of all kind).

**Comment by TheAssassin** [comment]
> was assuming that STDERR was for _errors_

Well, that nomenclature is as old as computing is. Any software that e.g., uses pipes uses `stderr` for all logging, since `stdout` is what will provide data for the pipe. Therefore almost all unix-ish tools use `stderr` for logging.

**Comment by MerryMurray** [comment]
HI
Well, few people will ever see this message since almost all appimages are gui apps.
Better to send this message to a log file and have something like:
   myappimage --image-status to display the info. 
and keep the commandline clean.
I ran the  lmms appimage without ever knowing i could create the .config and .home
folders since it was never launched from the terminal and not commented on their website.
If there is a .home and .config it is most likely (but not always) created by the user
anyway.
How does your typical user know of the existence of these files since they wont be
running them from a terminal?

Thanks

Murray

**Comment by probonopd** [comment]
@MerryMurray these options are for advanced users, and they are documented on https://docs.appimage.org/user-guide/portable-mode.html. Application authors are free to mention them in their documentation, too.

Changing the stderr output to a log file would not make them more discoverable, since "normal users" don't read log files either.

**Comment by TheAssassin** [comment]
The only reason not to print messages would be to not interfere with pipes etc. We are working on some better logging for libappimage, which might be introduced in appimagetool, too. We could add support for e.g., an environment variable to either hide or actually show the messages for debugging. I like the latter option more.

**Comment by probonopd** [comment]
What is wrong about sending the messages to stderr? What does it interfere with?

**Comment by probonopd** [comment]
Fight all configurability, imho It is the evil that makes software complex, cumbersome, and error-prone. The second we make messages optional, we get more complexity, questions, and unsatisfied users.

**Comment by MerryMurray** [comment]
dmesg | myappimage 2>/dev/null | myotherappimage 2>/dev/null | sort | grep...

**Comment by probonopd** [comment]
Once all informative output is in STDERR, this could be closed, correct?

**Comment by MerryMurray** [comment]
Well. Yes, fine with me but.. been thinking.
If the message is for advanced users:
1 normal users will never see it.
2 only advanced users who have already created .home.. will see it
  and do not need to.
3 advanced users who could use the feature but are not aware of it will never
  see the message because the message is only displayed after the folders 
  have been created. Message should be other way around - if they could
 have created a .home .config and did not let them know they can. That
 works much better since a normal run with .home has no output. Would
 this work for you?
4. dmesg | myappimage 2>/dev/null | myappmessage2 2>/dev/null | sort |
grep... is cumbersome.
5 if the purpose is discoverability this doesn't help imo, something else
needs to be done.

This a wonderful app and I'm enjoying learning how to create AppImages.

Best wishes

Murray

**Comment by TheAssassin** [comment]
@probonopd no, we haven't decided on whether to add some opt-in/opt-out for the log messages.

**Comment by probonopd** [comment]
My vote is on opt-out. If we should even allow this. Let's keep in mind that AppImage is designed for what `.app` bundles are on the Mac, not for some command line magic foo.

**Comment by TheAssassin** [comment]
I can come up with arguments for both opt-in and opt-out. None of the sides overweighs the onother, so I'm staying neutral here. As long as we just introduce an option to change the logging, I'm fine.

**Comment by probonopd** [comment]
By defaut make things in a way that satisfies 80% of the users and minimize support hassle, I'd say. The 20% experts know how to use Google :-)
